server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,178 @@
+
+   public ProtobufResource(InvocationHelper invocationHelper) {
+      super(invocationHelper);
+   }
+
+   @Override
+   public Invocations getInvocations() {
+      return new Invocations.Builder()
+            // Key related operations
+            .invocation().methods(GET).path("/v2/protobuf").handleWith(this::getSchemasNames)
+            .invocation().methods(PUT, POST).path("/v2/protobuf/{schemaName}").handleWith(this::addOrEditSchema)
+            .invocation().methods(GET).path("/v2/protobuf/{schemaName}").handleWith(this::getSchema)
+            .invocation().method(DELETE).path("/v2/protobuf/{schemaName}").handleWith(this::deleteSchema)
+            .create();
+   }
+
+   private CompletionStage<RestResponse> getSchemasNames(RestRequest request) {
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      List<ProtoSchema> protoSchemas = cache.entrySet().stream()
+            .filter(cacheEntry -> !((String)cacheEntry.getKey()).endsWith(ProtobufMetadataManager.ERRORS_KEY_SUFFIX))
+            .map(cacheEntry -> {
+               String error = (String) cache
+                     .get(cacheEntry.getKey() + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX);
+               ProtoSchema protoSchema = new ProtoSchema();
+               protoSchema.name = cacheEntry.getKey().toString();
+               protoSchema.validation = error == null ? "success" : error;
+               return protoSchema;
+            })
+            .collect(Collectors.toList());
+
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+      try {
+         responseBuilder
+               .status(OK)
+               .contentType(APPLICATION_JSON)
+               .entity(invocationHelper.getMapper().writeValueAsBytes(protoSchemas));
+      } catch (JsonProcessingException e) {
+         responseBuilder.status(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+      }
+
+      return CompletableFuture.completedFuture(responseBuilder.build());
+   }
+
+   private CompletionStage<RestResponse> addOrEditSchema(RestRequest request) {
+      String schemaName = checkMandatorySchemaName(request);
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      ContentSource contents = request.contents();
+      if (contents == null) throw new NoDataFoundException();
+      NettyRestResponse.Builder builder = new NettyRestResponse.Builder();
+      builder.status(HttpResponseStatus.OK);
+      return cache.putAsync(schemaName, contents.asString())
+            .thenCompose(r -> cache.getAsync(schemaName + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX))
+            .thenApply(validationError -> {
+               String validation = "success";
```

Comment:
No need to return a body if the validation is success: it should simply return code 204 (see above)